src/main/java/io/vertx/ext/mail/impl/SMTPSendMail.java

```diff
@@ -127,111 +105,210 @@ private void rcptToCmd() {
+  private Future<Boolean> sendMailEvenlope() {
+    Promise<Boolean> evenlopePromise = Promise.promise();
     try {
-      final String recipientAddr = recipientAddrs.get(i);
-      final String email;
-      if (EmailAddress.POSTMASTER.equalsIgnoreCase(recipientAddr)) {
-        email = recipientAddr;
+      if (checkSize()) {
+        final String mailFromLine = "MAIL FROM:<" + mailFromAddress() + ">" + sizeParameter();
+        final List<String> allRecipients = allRecipients();
+        if (config.isPipelining() && connection.getCapa().isCapaPipelining()) {
+          final List<String> groupCommands = new ArrayList<>();
+          groupCommands.add(mailFromLine);
+          groupCommands.addAll(allRecipients.stream().map(r -> "RCPT TO:<" + r + ">").collect(Collectors.toList()));
+          groupCommands.add("DATA");
+          connection.writeCommands(groupCommands, evenlopeResult -> {
+            if (groupCommands.size() != evenlopeResult.size()) {
+              evenlopePromise.fail("Sent " + groupCommands.size() + " commands, but got " + evenlopeResult.size() + " responses.");
+            } else {
+              // result follows the same order in the commands list
+              for (int i = 0; i < evenlopeResult.size(); i ++) {
+                String message = evenlopeResult.get(i);
+                if (log.isDebugEnabled()) {
+                  log.debug("S: " + message);
+                }
+                if (i == 0) {
+                  if (!StatusCode.isStatusOk(message)) {
+                    evenlopePromise.fail("sender address not accepted: " + message);
+                    return;
+                  }
+                } else if (i < evenlopeResult.size() - 1) {
+                  if (StatusCode.isStatusOk(message)) {
+                    mailResult.getRecipients().add(allRecipients.get(i - 1));
+                  } else {
+                    if (config.isAllowRcptErrors()) {
+                      log.warn("recipient address not accepted, continuing: " + message);
+                    } else {
+                      evenlopePromise.fail("recipient address not accepted: " + message);
+                      return;
+                    }
+                  }
+                } else {
+                  // DATA result
+                  if (StatusCode.isStatusOk(message)) {
+                    if (mailResult.getRecipients().size() == 0) {
+                      // send dot only
+                      evenlopePromise.complete(false);
+                      return;
+                    }
+                  } else {
+                    evenlopePromise.fail("DATA command not accepted: " + message);
+                    return;
+                  }
+                }
+              }
+              evenlopePromise.complete(true);
+            }
+          });
+        } else {
+          // sent line by line
+          Future<Void> future = sendMailFrom(mailFromLine);
+          for (String email: allRecipients) {
```

Comment:
doing this introduces delay, why not batch all in a single write ?

Reply:
These are multiple SMTP commands(`RCPT TO:`) which means multiple `command <-> reply` conversations because server does not support `PIPELINING`. It needs client to make a decision on each reply.

Reply:
ok, I didn't spot them